src/core/Optional.h

```diff
@@ -0,0 +1,86 @@
+/*
+ *  Copyright (C) 2017 KeePassXC Team <team@keepassxc.org>
+ *
+ *  This program is free software: you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation, either version 2 or (at your option)
+ *  version 3 of the License.
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#ifndef OPTIONAL_H
+#define OPTIONAL_H
+
+/*
+ * This utility class is for providing basic support for an option type.
+ * It can be replaced by std::optional (C++17) or
+ * std::experimental::optional (C++11) when they become fully supported
+ * by all the main compiler toolchains.
+ */
+
+template <typename T> class Optional
```

Comment:
Do we really need our own Optional implementation? Would the same be possible with QVariant?

Reply:
I think we can make this change to reduce the code base size, as the variant type can serve as an optional.

Reply:
Just just QVariant::isNull() or QVariant::isValid().

Reply:
I'm almost done making this change, I decided for `QVariant::isNull`; as I understand that`QVariant::isValid` is used for detecting when a type not defined for the variant (i.e. not registered with Q_DECLARE_METATYPE) is used.